### PR TITLE
feat (l2): hard fork

### PR DESCRIPTION
**Pull request by faultytolly**
## Description

----

Closes #XXX

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

----

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/sequencers/keeper/msg_server.go [hunk 1]
```diff
@@ -113,3 +118,84 @@ func (m msgServer) UpsertSequencer(goCtx context.Context, msg *types.ConsensusMs
...
+const BumpSequence = 1_000_000_000
+
+func (m msgServer) BumpAccountSequences(goCtx context.Context, msg *types.MsgBumpAccountSequences) (*types.MsgBumpAccountSequencesResponse, error) {
+	if msg.Authority != m.authority {
+		return nil, sdkerrors.ErrorInvalidSigner.Wrapf("only an authorized actor can bump account sequences")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	var allErrors error
```
Comment (danwt) [inline x/sequencers/keeper/msg_server.go:141]: This is weird, you're making a stack of errors
You should only call Join once

File: x/sequencers/keeper/msg_server.go [hunk 2]
```diff
@@ -113,3 +118,84 @@ func (m msgServer) UpsertSequencer(goCtx context.Context, msg *types.ConsensusMs
...
+				toBump, err := f(accType, account)
+				if err != nil {
+					allErrors = errors.Join(allErrors, fmt.Errorf("filter account: %w", err))
+					return false
+				}
+				if toBump {
+					err := m.bumpAccountSequence(ctx, account)
+					allErrors = errors.Join(allErrors, err)
+					break
+				}
```
Comment (danwt) [inline x/sequencers/keeper/msg_server.go:161]: Can refac to only call bump once

**Comment by danwt** [comment]
+1 height is a blocker

**Comment by danwt** [comment]
@srene  

> why do we have two separate consensus messages for the upgrade? bump account sequencers + updateDRS? couldnt we do it with a unique upgradedrs message and include the drs update version and bumpsequences logic in the necssary upgrade handler?

Hey sergi

-Not all upgrades will require sequence bump
-Sometimes you want to sequence bump without upgrade

File: x/sequencers/keeper/msg_server.go [hunk 3]
```diff
@@ -113,3 +118,97 @@ func (m msgServer) UpsertSequencer(goCtx context.Context, msg *types.ConsensusMs
...
+
+func (m msgServer) BumpAccountSequences(goCtx context.Context, msg *types.MsgBumpAccountSequences) (*types.MsgBumpAccountSequencesResponse, error) {
+	if msg.Authority != m.authority {
+		return nil, sdkerrors.ErrorInvalidSigner.Wrapf("only an authorized actor can bump account sequences")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	var allErrors error
+	m.accountKeeper.IterateAccounts(ctx, func(account authtypes.AccountI) bool {
```
Comment (mtsitrin) [inline x/sequencers/keeper/msg_server.go:142]: why u not breaking with error on the first error?

Comment (danwt) [reply x/sequencers/keeper/msg_server.go:142]: fixed

File: x/sequencers/keeper/msg_server.go [hunk 4]
```diff
@@ -113,3 +118,91 @@ func (m msgServer) UpsertSequencer(goCtx context.Context, msg *types.ConsensusMs
...
+//   - bool: true if the version was updated, false if no update was needed (versions were identical)
+func (m msgServer) updateDrsVersion(ctx sdk.Context, newVersion uint64) bool {
+	currentParams := m.rollapParamsKeeper.GetParams(ctx)
+	if currentParams.DrsVersion == uint32(newVersion) {
+		return false
+	}
+
+	currentParams.DrsVersion = uint32(newVersion)
+	m.rollapParamsKeeper.SetParams(ctx, currentParams)
+
```
Comment (danwt) [inline x/sequencers/keeper/msg_server.go:206]: still not sure about this but nvm

**Review by danwt — APPROVED** [review]

File: x/sequencers/keeper/msg_server.go [hunk 5]
```diff
@@ -113,3 +118,91 @@ func (m msgServer) UpsertSequencer(goCtx context.Context, msg *types.ConsensusMs
...
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	var err error
+	m.accountKeeper.IterateAccounts(ctx, func(account authtypes.AccountI) bool {
+		accType := proto.MessageName(account)
+		_, ok := handleAccounts[accType]
+		i := 0
+		for !ok && i < len(m.accountBumpFilters) {
```
Comment (mtsitrin) [inline x/sequencers/keeper/msg_server.go:146]: if a filter returns ok=true, the loop still continues instead of breaking and calling bumpAccountSequence

Comment (mtsitrin) [inline x/sequencers/keeper/msg_server.go:146]: why not using ` range m.accountBumpFilters`

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: x/sequencers/keeper/msg_server.go (hunk 5 above)
Comment (danwt) [reply x/sequencers/keeper/msg_server.go:146]: the loop is `for !ok..`

**Pull request merged**